Implement PDO based DB drivers for MySQL and PostgreSQL

euantorano opened the pull request:
This PR adds two new database drivers for MySQL and PostgreSQL that make use of [PDO](https://www.php.net/manual/en/book.pdo.php).

It also amends the SQLite driver to not just die if the database file cannot be opened.

It also fixes #4192.

euantorano committed:
dbpdoEngine enhancements

Enahnce the dbpdoEngine class in a few ways:

- Disable emulated prepares.
- Rather than dying when failing to connect, throw an exception.
- Make some properties private where possible.
- Add comments to properties.
- Add `query_prepared` method.
- Use object hash to identify queries, rather than assigning an ID property.
- Default name of query for `insert_id` to `NULL` rather than an empty string. This matches the underlying function.
- Set visibility for functions.

euantorano committed:
Working on PDO versions of MySQL and PostgreSQL DB

euantorano committed:
Implement further postgres PDO methods.

euantorano committed:
Fix spelling, implement some more methods

euantorano committed:
Improve SELECT query detection.

Implement `insert_query`, `insert_query_multiple`, `update_query`.
Use heredoc to format HTML for query explanations, which makes the HTML more readable.

euantorano committed:
Implement `show_fields_from`.

euantorano committed:
Implement more DB methods for postgres PDO driver

euantorano committed:
Move database methods for Postgres PDO connector

euantorano committed:
Start working in integrating Postgres PDO driver

Moved some methods to the postgres specific PDO class as they have different requirements (such as rewriting LIMITs).

euantorano committed:
Handle escaping binary and decoding binary

Escaping is handled slightly differently to the standard pgsql driver

euantorano committed:
Implement missing methods. Handl DB type property.

euantorano committed:
Starting work on mySQL PDO driver.

euantorano committed:
Implement more mYSQL PDO methods

euantorano committed:
Fix #4192 - set default when modifying columns

euantorano committed:
Drop the query_prepared function for now.

euantorano committed:
Handle exception opening SQLite DB.

euantorano commented on inc/AbstractPdoDbDriver.php:
```diff
@@ -0,0 +1,713 @@
+	{
+		// first, check for an IPv6 address - IPv6 addresses always start with `[`
+		$openingSquareBracket = strpos($hostname, '[');
+		if ($openingSquareBracket === 0) {
```
Note: This logic is not yet tested. I believe that other drivers doing much the same thing would never have worked with IPv6 addressed DB hosts...

euantorano commented on inc/AbstractPdoDbDriver.php:
```diff
@@ -0,0 +1,713 @@
+			return false;
+		}
+
+		if (preg_match('/^\\s*SELECT\\b/i', $query->queryString) === 1) {
```
Note: using `num_rows` with PDO is not optimal at all. We need to review usages within the core and see when/why we use `num-rows` and if they can be replaced so that we don't hit this path too often.

This works better with the old `mysql_` type functions as those functions buffer results internally, which PDO does not do.

euantorano commented on inc/db_pgsql_pdo.php:
```diff
@@ -0,0 +1,684 @@
+		return $fieldstring;
+	}
+
+	public function __set($name, $value)
```
Note: This special handling is important to ensure compatibility.

euantorano committed:
Implement all MySQL functions.

euantorano committed:
Handle mysql_pdo DB type.

euantorano commented:
I haven't yet tested the MySQL PDO driver, but have tested the Postgres one. I'd like to land this in 1.8.25 but not make too much noise about it. Then during the cycle for 1.8.26 I plan to add some more methods to all of the DB drivers to support prepared statements (which will be emulated for some drivers).

euantorano committed:
Prefer string inteprolation over concatenation.

euantorano committed:
Check for MySQL in supported PDO drivers.

euantorano committed:
Make explain property public and fix formatting of explain. Fix list_tables method.

euantorano commented:
I've now done some testing of this with MySQL too and ran through an install and a few key pages.

I'd very much appreciate it if others could test this PR. Upgraded boards can opt-in to the PDO based drivers by modifying `inc/config.php` to alter the DB engine to one of the following options:

```php
$config['database']['type'] = 'mysql_pdo';
```

```php
$config['database']['type'] = 'pgsql_pdo';
```

Once altered, there shouldn't be any visible change if all goes well.

euantorano committed:
Close cursor for query in free_result.

lairdshaw commented:
> I'd very much appreciate it if others could test this PR.

I haven't done any systematic testing, but have been using the `mysql_pdo` driver for about two weeks on my local dev board whilst working on a plugin, and have had no problems so far. Seems good.

awslabspl commented:
> > I'd very much appreciate it if others could test this PR.
> 
> I haven't done any systematic testing, but have been using the `mysql_pdo` driver for about two weeks on my local dev board whilst working on a plugin, and have had no problems so far. Seems good.

Second this. Also no problems.

euantorano commented:
Awesome, thanks Laird!

On Sat, 13 Feb 2021, at 11:36, Laird wrote:
> 
> 
> > I'd very much appreciate it if others could test this PR.
> 
> I haven't done any systematic testing, but have been using the 
> `mysql_pdo` driver for about two weeks on my local dev board whilst 
> working on a plugin, and have had no problems so far. Seems good.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mybb/mybb/pull/4193#issuecomment-778604182>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAFW24L2WITDRPAKKIJVXS3S6ZP2TANCNFSM4UZXSIOA>.
>

dvz commented on inc/AbstractPdoDbDriver.php:
```diff
@@ -0,0 +1,717 @@
+				$portSeparator = strpos($hostname, ':', $closingSquareBracket);
+
+				// there is no port specified
+				if ($portSeparator === false && strlen($hostname) > $portSeparator + 1) {
```
Does the second condition make sense (also on line 318)?

euantorano replied:
Looking at it again, no I don't think it does.

This is the kind of thing that would be best tested by some unit tests, if we had any...

euantorano committed:
Update function to parse hostname

Previous incarnation missed the closing `]` for IPv6 addresses, and had some questionable logic.

dvz commented on inc/db_pgsql.php:
```diff
@@ -451,6 +451,8 @@ function fetch_field($query, $field, $row=false)
 			return pg_fetch_result($query, $row, $field);
 		}
+
+		return pg_fetch_result($query, $row, $field);
```
Unreachable statement

euantorano commented:
Seems some changes were missing for this. I think I’d simply dropped the else but must not have staged it. Will fix later. 

On Thu, 25 Mar 2021, at 20:28, dvz wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> In inc/db_pgsql.php 
> <https://github.com/mybb/mybb/pull/4193#discussion_r601817233>:
> 
> > @@ -451,6 +451,8 @@ function fetch_field($query, $field, $row=false)
>  		{
>  			return pg_fetch_result($query, $row, $field);
>  		}
> +
> +		return pg_fetch_result($query, $row, $field);
> Unreachable statement
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mybb/mybb/pull/4193#pullrequestreview-621550463>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAFW24JGJ6VIHNWDHYZB2ULTFOMIPANCNFSM4UZXSIOA>.
>

euantorano committed:
Fix unreachable statement

euantorano commented:
I'm going ahead and merging this, it's had a fair bit of testing now I believe.
